File: bindings/typescript/test/mgep.test.ts
// Framework-free tests for MGEP TypeScript bindings. Runnable with:
//   node --experimental-strip-types test/mgep.test.ts   (Node 22+)
//   deno test test/mgep.test.ts                          (Deno)
//   bun test test/mgep.test.ts                           (Bun)

import {
  MAGIC,
  VERSION,
  FULL_HEADER_SIZE,
  CORE_BLOCK_OFFSET,
  DECIMAL_NULL,
  Side,
  OrderType,
  TimeInForce,
  SCHEMA,
  MSG,
  NewOrderSingleSize,
  ExecutionReportSize,
  BusinessRejectSize,
  BookSnapshotBeginSize,
  BookSnapshotLevelSize,
  BookSnapshotEndSize,
  ClockStatusSize,
  encodeDecimal,
  decodeDecimal,
  encodeNewOrder,
  decodeHeader,
  decodeNewOrder,
} from "../src/mgep.ts";

let passed = 0;
let failed = 0;

function assertEq<T>(a: T, b: T, name: string): void {
  if (a === b) {
    passed++;
    console.log(`PASS: ${name}`);
  } else {
    failed++;
    console.log(`FAIL: ${name} — ${String(a)} !== ${String(b)}`);
  }
}

function assertClose(a: number, b: number, tol: number, name: string): void {
  if (Math.abs(a - b) < tol) {
    passed++;
    console.log(`PASS: ${name}`);
  } else {
    failed++;
    console.log(`FAIL: ${name} — ${a} !≈ ${b}`);
  }
}

function assertTruthy(cond: unknown, name: string): void {
  if (cond) {
    passed++;
    console.log(`PASS: ${name}`);
  } else {
    failed++;
    console.log(`FAIL: ${name}`);
  }
}

// ─── Size checks ────────────────────────────────────────────

assertEq(NewOrderSingleSize, 48, "NewOrderSingle size");
assertEq(ExecutionReportSize, 88, "ExecutionReport size");
assertEq(BusinessRejectSize, 16, "BusinessReject size");
assertEq(BookSnapshotBeginSize, 40, "BookSnapshotBegin size");
assertEq(BookSnapshotLevelSize, 40, "BookSnapshotLevel size");
assertEq(BookSnapshotEndSize, 32, "BookSnapshotEnd size");
assertEq(ClockStatusSize, 40, "ClockStatus size");

// ─── Decimal roundtrip ──────────────────────────────────────

assertEq(encodeDecimal(150.25), 15_025_000_000n, "decimal encode");
assertClose(decodeDecimal(15_025_000_000n) as number, 150.25, 1e-9, "decimal decode");
assertEq(encodeDecimal(null), DECIMAL_NULL, "null sentinel");
assertEq(decodeDecimal(DECIMAL_NULL), null, "null roundtrip");

// ─── New order encode + decode ──────────────────────────────

const bytes = encodeNewOrder({
  orderId: 42n,
  clientOrderId: 99n,
  instrumentId: 7,
  side: Side.Buy,
  orderType: OrderType.Limit,
  price: 150.25,
  quantity: 100.0,
  timeInForce: TimeInForce.Day,
  senderCompId: 1,
  sequenceNum: 1n,
  correlationId: 0n,
});
assertEq(bytes.length, FULL_HEADER_SIZE + NewOrderSingleSize, "encoded size");

const hdr = decodeHeader(bytes);
assertTruthy(hdr !== null && hdr.valid, "header valid");
assertEq(hdr!.magic, MAGIC, "magic");
assertEq(hdr!.version, VERSION, "version");
assertEq(hdr!.schemaId, SCHEMA.Trading, "schema_id");
assertEq(hdr!.messageType, MSG.NewOrder, "message_type");

const order = decodeNewOrder(bytes);
assertEq(order.orderId, 42n, "order_id");
assertEq(order.clientOrderId, 99n, "client_order_id");
assertEq(order.instrumentId, 7, "instrument_id");
assertEq(order.side, Side.Buy, "side");
assertEq(order.orderType, OrderType.Limit, "order_type");
assertClose(decodeDecimal(order.price) as number, 150.25, 1e-9, "price");
assertClose(decodeDecimal(order.quantity) as number, 100.0, 1e-9, "quantity");
assertEq(order.stopPrice, DECIMAL_NULL, "stop_price null");

// ─── Header validation ──────────────────────────────────────

assertEq(decodeHeader(new Uint8Array(16)), null, "short buffer rejected");
const zero = new Uint8Array(FULL_HEADER_SIZE);
const zeroHdr = decodeHeader(zero);
assertTruthy(zeroHdr !== null && !zeroHdr.valid, "zero-magic invalid");

// ─── Wire-offset verification ───────────────────────────────

// Byte 0..2 should be MAGIC little-endian: 0x4D, 0x47.
assertEq(bytes[0], 0x4d, "magic byte 0");
assertEq(bytes[1], 0x47, "magic byte 1");

// Offset 32 (CORE_BLOCK_OFFSET) is start of core. order_id=42 → 0x2A at offset 32.
assertEq(bytes[CORE_BLOCK_OFFSET], 0x2a, "order_id byte 0");

console.log(`\nResults: ${passed} passed, ${failed} failed`);
if (failed > 0) {
  // eslint-disable-next-line no-undef
  (globalThis as { process?: { exit: (code: number) => void } }).process?.exit(1);
}

File: bindings/java/src/test/java/com/mgep/MgepTest.java
package com.mgep;

/**
 * Minimal test harness — intentionally framework-free so the bindings
 * have no build dependencies beyond a JDK. A full project would swap
 * this for JUnit 5.
 *
 * Run with: javac -d out src/main/java/com/mgep/*.java src/test/java/com/mgep/*.java && java -cp out com.mgep.MgepTest
 */
public final class MgepTest {
    private static int passed = 0;
    private static int failed = 0;

    public static void main(String[] args) {
        testSizes();
        testNewOrderRoundtrip();
        testDecimalEncoding();
        testHeaderDecode();

        System.out.printf("%nResults: %d passed, %d failed%n", passed, failed);
        if (failed > 0) System.exit(1);
    }

    private static void testSizes() {
        assertEq(Mgep.NewOrderSingle.SIZE, 48, "NewOrderSingle size");
        assertEq(Mgep.ExecutionReport.SIZE, 88, "ExecutionReport size");
        assertEq(Mgep.BusinessReject.SIZE, 16, "BusinessReject size");
        assertEq(Mgep.BookSnapshotBegin.SIZE, 40, "BookSnapshotBegin size");
        assertEq(Mgep.BookSnapshotLevel.SIZE, 40, "BookSnapshotLevel size");
        assertEq(Mgep.BookSnapshotEnd.SIZE, 32, "BookSnapshotEnd size");
        assertEq(Mgep.ClockStatus.SIZE, 40, "ClockStatus size");
    }

    private static void testNewOrderRoundtrip() {
        byte[] bytes = Mgep.encodeNewOrder(
            42L, 99L, 7,
            Mgep.SIDE_BUY, Mgep.ORDER_TYPE_LIMIT,
            150.25, 100.0,
            Mgep.TIF_DAY, 1, 1L, 0L);

        assertEq(bytes.length, Mgep.FULL_HEADER_SIZE + Mgep.NewOrderSingle.SIZE,
                "encoded size");

        Mgep.FullHeader hdr = Mgep.FullHeader.decode(bytes);
        assertTrue(hdr != null && hdr.isValid(), "header valid");
        assertEq(hdr.schemaId, Mgep.SCHEMA_TRADING, "schema_id");
        assertEq(hdr.messageType, Mgep.MSG_NEW_ORDER, "message_type");

        Mgep.NewOrderSingle order = Mgep.NewOrderSingle.decode(bytes);
        assertEq(order.orderId, 42L, "order_id");
        assertEq(order.clientOrderId, 99L, "client_order_id");
        assertEq(order.instrumentId, 7, "instrument_id");
        assertEq(order.side, Mgep.SIDE_BUY, "side");
        assertCloseTo(order.priceAsDouble(), 150.25, 1e-6, "price");
        assertCloseTo(order.quantityAsDouble(), 100.0, 1e-6, "quantity");
    }

    private static void testDecimalEncoding() {
        long enc = Mgep.encodeDecimal(150.25);
        assertEq(enc, 15_025_000_000L, "decimal encode");
        double dec = Mgep.decodeDecimal(enc);
        assertCloseTo(dec, 150.25, 1e-9, "decimal roundtrip");
        // NULL sentinel
        assertEq(Mgep.encodeDecimal(Double.NaN), Mgep.DECIMAL_NULL, "NaN → DECIMAL_NULL");
    }

    private static void testHeaderDecode() {
        // Malformed buffer → null
        assertTrue(Mgep.FullHeader.decode(new byte[16]) == null, "short buffer rejected");
        // Zero buffer → invalid magic
        Mgep.FullHeader h = Mgep.FullHeader.decode(new byte[Mgep.FULL_HEADER_SIZE]);
        assertTrue(h != null && !h.isValid(), "zero magic invalid");
    }

    // ─── Primitive assertions ──────────────────────────────

    private static void assertEq(long a, long b, String msg) {
        if (a == b) { passed++; System.out.println("PASS: " + msg); }
        else { failed++; System.out.printf("FAIL: %s — %d != %d%n", msg, a, b); }
    }

    private static void assertCloseTo(double a, double b, double tol, String msg) {
        if (Math.abs(a - b) < tol) { passed++; System.out.println("PASS: " + msg); }
        else { failed++; System.out.printf("FAIL: %s — %f !≈ %f%n", msg, a, b); }
    }

    private static void assertTrue(boolean cond, String msg) {
        if (cond) { passed++; System.out.println("PASS: " + msg); }
        else { failed++; System.out.println("FAIL: " + msg); }
    }
}
